File: tag.py
import re
import csv
import StringIO


class Tag:
    tagstring = 'tag'
    content = ''
    tag = ''

    def __init__(self, tagstring):
        self.tagstring = tagstring

    def processTags(self, content):
        newcontent = content
        self.content = content
        tagstringoffset = len(self.tagstring) + 2

        pattern = re.compile('(\[' + self.tagstring + ':[a-zA-Z]*\([^\[\]]*\)\])')
        for tag in re.findall(pattern, content):
            self.tag = tag
            tagpattern = re.compile(".*:(.*)\((.*)\)")
            matches = re.search(tagpattern, tag)
            funcstr = matches.groups(0)[0]
            paramstring = matches.groups(0)[1]
            reader = csv.reader(StringIO.StringIO(paramstring), delimiter=',', skipinitialspace=True)
            params = []
            for row in reader:
                params = row
            try:

                func = getattr(self, funcstr)
            except AttributeError:
                0  # do nothing
            else:

                newcontent = func(params)

            self.content = newcontent
        return newcontent

    def replace(self, newstring):
        return self.content.replace(self.tag, newstring, 1)

    # *************************************************
    #** Example Tag Functions ************************
    #*************************************************
    # Notes:
    # Very simple example function
    #
    def helloWorld(self, args):
        return self.replace("Hello World");

    #
    # This would replace the tag [tag:helloWorld()] with Hello World
    #

    #
    # if you want parameters passed into the function, use the $args array parameter
    # simple example with 1 arg
    #
    def displayLine(self, args):
        line = "<div style='width:%s;height:0px;border-bottom:#000 1px solid;' ></div>" % args[0]
        return self.replace(line)

    #
    # The tag would look like: [tag:displayLine(25px)] or [tag:displayLine("25px")]
    #
    # simple example with 2 args
    #
    def displaySquare(self, args):
        square = "<div style='width:%s;height:%s;border:#000 1px solid;' ></div>" % (args[0], args[1])
        return self.replace(square)

    #
    # The tag would look like: [tag:displaySquare(100px,100px)] or [tag:displaySquare("100px","100px")]
    #

    def todaysDate(self, args):
        import time

        return self.replace(time.strftime('%A %B %d, %Y'))

    def shortDate(self, args):
        import time

        return self.replace(time.strftime("%m/%d/%Y"))

    def comment(self, args):
        return self.replace('')

    def youtube(self, args):
        url = args[0]
        width = 560
        height = 315
        if (len(args) == 3):
            width = args[1]
            height = args[2]

        from urlparse import urlparse

        o = urlparse(url)
        path = o.path

        vidcontent = '<iframe width="' + width + '" height="' + height + '" src="//www.youtube.com/embed' + path + '" frameborder="0" allowfullscreen></iframe>'

        return self.replace(vidcontent)

File: Tag.php
<?php
namespace Negativenull\Cbatags;


class Tag {

	private $tagstring;
	private $cond;
	private $content;
	private $tag;
	private $variables;
	function __construct($tagstring='tag') {
		$this->tagstring=$tagstring;
		$this->cond=true;
		$this->variables = array();
	}

	function processTags($content) {
		$newContent = $content;
		$this->content = $content;
		$tagstringoffset = strlen($this->tagstring)+2;
		$offset=0;
		
		$pattern = '|\['.$this->tagstring.':.*(.*)\]|';
		
		preg_match_all($pattern, $content, $matches);
		
		foreach($matches[0] as $tag) {
			$this->tag = $tag;
			preg_match("/.*:(.*)\((.*)\)/", $tag, $output_array);
			$func = $output_array[1];
			$parmsstr = $output_array[2];
			$parms = str_getcsv($parmsstr);

			if((method_exists($this, $func) && $this->cond) || $func=='endcond') {
				$newContent = $this->$func($parms);
				$alterContent=true;
			}
			if(!$this->cond) {
				$endcondpos=strpos ( $newContent , "[".$this->tagstring.":endcond()]",$offset);
				$newContent = substr_replace ( $newContent, "" , $start, $endcondpos-4);
			}
			$this->content = $newContent;
		}
		return $newContent;
		
		
	}

	
// *****************************************************************************************************************************************
// ** Special Helper Functions *************************************************************************************************************
// *****************************************************************************************************************************************
	private function replace($newstring) {
		return $this->str_replace_once($this->tag, $newstring, $this->content);
	}
	private function str_replace_once($needle, $replace, $haystack) {
	   $pos = strpos($haystack, $needle);
	   if ($pos === false) {
		   // Nothing found
		   return $haystack;
	   }
	   
	   return substr_replace($haystack, $replace, $pos, strlen($needle));
	}
	



//*****************************************************************************************************************************************
//** Example Tag Functions ****************************************************************************************************************
//*****************************************************************************************************************************************
// Notes:
// Very simple example function
// 
	private function helloWorld($args=array()) {
		return $this->replace("Hello World");
	}
// 
// This would replace the tag [tag:helloWorld()] with Hello World
//

//
// if you want parameters passed into the function, use the $args array parameter
// simple example with 1 arg
//
	private function displayLine($args=array()) {
		$line = "<div style='width:".$args[0].";height:0px;border-bottom:#000 1px solid;' ></div>";
		return $this->replace($line);
	}
	//
// The tag would look like: [tag:displayLine(25px)] or [tag:displayLine("25px")]
//
// simple example with 2 args
//
	private function displaySquare($args=array()) {
		$square = "<div style='width:".$args[0].";height:".$args[1].";border:#000 1px solid;' ></div>";
		return $this->replace($square);
	}
//
// The tag would look like: [tag:displaySquare(100px,100px)] or [tag:displaySquare("100px","100px")]
//

//*****************************************************************************************************************************************
//** Example Tag Functions ****************************************************************************************************************
//*****************************************************************************************************************************************
	

	
	private function todaysDate($args=array()) {
		return $this->replace(date('l F j, Y'));
	}
	private function shortDate($args=array()) {
		return $this->replace(date('n/j/y'));
	}

	private function comment($args=array()) {
		return $this->replace('');
	}
	

	//example usage:
	// [tag:youtube("http://youtu.be/lxNORk0vKd0",300,200)]
	// [tag:youtube("http://youtu.be/lxNORk0vKd0")]
	private function youtube($args=array()) {
		$videostring=$args[0];
		$width=$args[1];
		$height=$args[2];
		if(empty($width)) $width=560;
		if(empty($height)) $height=315;
		$url = parse_url($videostring);
		$path = $url['path'];
		
		list($var,$value)=explode("=",$querystring);
	
		$vidcontent = '<iframe width="'.$width.'" height="'.$height.'" src="//www.youtube.com/embed'.$path.'" frameborder="0" allowfullscreen></iframe>';

		return $this->replace($vidcontent);
		
	}
	
	
	
//***********************************  Condictionals and variables (not quite working yet)
	private function assignvar($args=array()) {
		$variable = $args[0];
		$value = $args[1];
		$this->variables[$variable] = $value;
		return $this->replace('');
	}
	private function assignvartovar($args=array()) {
		$variable1 = $args[0];
		$variable2 = $args[1];
		$this->variables[$variable1] = $this->variables[$variable2];
		return $this->replace("");
	}
	private function getvar($args=array()) {
		$variable = $args[0];
		return $this->replace($this->variables[$variable]);
	}
	private function variable($args=array()) {
		$variable = $args[0];
		$value = $args[1];
		if(empty($value)) {
			return $this->replace($this->variables[$variable]);
		} else {
			$this->variables[$variable] = $value;
			return $this->replace("");
		}
	}
	private function add($args=array()) {
		$variable = $args[0];
		$value = $args[1];
		$this->variables[$variable]+=$value;
		
		return $this->replace("");
	}
	private function subtract($args=array()) {
		$variable = $args[0];
		$value = $args[1];
		$this->variables[$variable]-=$value;

		return $this->replace("");
	}
	private function multiply($args=array()) {
		$variable = $args[0];
		$value = $args[1];
		$this->variables[$variable]*=$value;

		return $this->replace("");
	}
	private function divide($content, $args=array()) {
		$variable = $args[0];
		$value = $args[1];
		$this->variables[$variable]/=$value;

		return $this->replace("");
	}
	
	
	private function cond($args=array()) {
		$var1 = $args[0];
		$var2 = $args[2];
		$operator = $args[1];

		if(!is_numeric($var1)) {
			$var1 = $this->variables[$var1];
		}
		if(!is_numeric($var2)) {
			$var2 = $this->variables[$var2];
		}

		switch($operator) {
			case "=":
				if((float)$var1===(float)$var2) {
					$this->cond=true;
				} else {
					$this->cond=false;
				}
				break;
			case "&lt;":
			case "<":
				if($var1<$var2)  {
					$this->cond=true;
				} else {
					$this->cond=false;
				}
				break;
			case "&gt;":
			case ">":
				if($var1>$var2)  {
					$this->cond=true;
				} else {
					$this->cond=false;
				}
				break;
			case "&lt;=":
			case "<=":
				if($var1<=$var2)  {
					$this->cond=true;
				} else {
					$this->cond=false;
				}
				break;
			case "&gt;=":
			case ">=":
				if($var1>=$var2)  {
					$this->cond=true;
				} else {
					$this->cond=false;
				}
				break;
			case "!=":
				if($var1!==$var2)  {
					$this->cond=true;
				} else {
					$this->cond=false;
				}
				break;
		}
		return $this->replace("");

	}

	private function endcond($args=array()) {
		$this->cond=true;
		return $this->replace("");
	}

}
?>
